# call-NonCanonicalPeptides

- [call-NonCanonicalPeptides](#call-noncanonicalpeptides)
  - [Overview](#overview)
  - [How To Run](#how-to-run)
  - [Flow Diagram](#flow-diagram)
  - [Input CSV](#input-csv)
  - [Config](#config)
  - [Outputs](#outputs)
  - [License](#license)
## Overview

This pipeline takes genomic and transcriptomic variation data such as SNP, INDEL, and Fusion, and call variant peptides from them using the graph-based algorithm [moPepGen](https://github.com/uclahs-cds/private-moPepGen)

---

## How To Run

1. Create sample specifc config file using the [template](config/template.config)

2. Create a sample specific input CSV file using the [tempalte](input/../inputs/input.csv)

3. To run on UCSL-CDS' Azure clusters, see the submission script, [here](https://github.com/uclahs-cds/tool-submit-nf), to submit it. For general usage, launch with the comamnd below:

```
nextflow run paht/to/pipeline-call-NoncanonicalPeptide/main.nf -c sample.config
```

> :warning: This pipeline should only run one sample at a time. The input CSV file should only contain one sample for label free proteomics. For isobaric labeled (e.g., TMT) proteomics, the CSV file should contain all samples within a TMT plex.

---

## Flow Diagram

![flow-chart](img/diagram.drawio.svg?raw=true)

---

## Input CSV

The input CSV file must contain the fields listed below. See example [here](input/../inputs/input.csv)

| Field name | Required | Description |
| ---------- | -------- | ----------- |
| sample_name | yes | Sample name. |
| software | yes | The software used to call this variant. Must come from VEP, STAR-Fusion, rMATS, CIRCexplorer, and REDItools. |
| alt_splic_type | no | Alternative splicing type. Required for rMATS. Must come from SE, A5SS, A3SS, MXE, and RI. |
| source | yes | Source of the variant. For example, gSNP, sSNV, Fusion, circRNA, etc |
| path | yes | Path to the variant file. |

---

## Config

| Field name | Required | Description |
| ---------- | -------- | ----------- |
| `output_dir` | yes | Output directory. |
| `temp_dir` | yes | Directory for intermediate files. This can be set to /scratch. Default to the launchDir. |
| `index_dir` | yes | Path the the genome index directory, generated by `moPepGen generateIndex`. |
| `min_est_j` | no | Minimal estimate junction reads. Used by `moPepGen parseSTARFusion`. |
| `transcript_id_column` | no | Column index for transcript ID. Used by `moPepGen parseREDItools`. |
| `min_read_number` | no | Minimal read number. Used by `moPepGen parseCIRCexplorer`. |
| `min_fpb_circ` | no | Minimal FPB (fragment per billion) of the circRNA. Used by `moPepGen parseCIRCexplorer`. |
| `min_circ_score` | no | Minimal circRNA score. Used by `moPepGen parseCIRCexplorer`. |
| `miscleavage` | no | Maximal miscleavages allowed. Used by `moPepGen callVariant`. |
| `min_mw` | no | Minimal molecular weight of the variant peptide to keep. Used by `moPepGen callVariant` |
| `min_length` | no | Minimal length of the variant peptide to keep. Used by `moPepGen callVariant` |
| `max_length` | no | Maximal length of the variant peptide to keep. Used by `moPepGen callVariant` |
| `split_database` | no | Boolean value. When false, the `split_databse` step is skipped. Defaults to `true`. |
| `noncoding_fasta` | no | Path to the noncoding peptide FASTA, called by `moPepGen callNoncoding` |
| `order_source` | no | Required when `split_database` is true. The order of variant sources.  Must be a string seperated by comma. |
| `group_source` | no | This is used when you want to group certain sources. For example, 'germline:gSNP,gINDEL' |
| `max_source_groups` | no | Maximal number of different source groups considered together. Defaults to 2. |
| `additional_split` | no | For peptides that were not already split into FASTAs up to `max_source_groups`, those involving the following source will be split into additional FASTAs with decreasing priority |

---

## Outputs

 Output and Output Parameter/Flag | Description |
| ------------ | ------------------------ |
| <sample_name>_<source>_<software>.gvf | Intermediate GVF files. |
| <sample_name>_variantPeptide.fasta | The complete variant peptide FASTA file. |
| split/<sample_name>_<source>_<software>.fasta | Splited database FASTA files can be used for multi-step library search and FDR calculation. |

---

## License

Author: Chenghao Zhu (ChenghaoZhu@mednet.ucla.edu)

pipeline-call-NoncanonicalPeptide is licensed under the GNU General Public License version 2. See the file LICENSE for the terms of the GNU GPL license.

pipeline-call-NoncanonicalPeptide is a nextflow pipeline to call non-canonical peptides as custom databases for proteogenomic analysis.

Copyright (C) 2021 University of California Los Angeles ("Boutros Lab") All rights reserved.

This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
